packages/lib/contracts/upgradeability/Proxy.sol

```diff
@@ -28,23 +28,12 @@ contract Proxy {
    * @param implementation Address to delegate.
    */
   function _delegate(address implementation) internal {
+    (bool success, bytes memory data) = implementation.delegatecall(msg.data);
     assembly {
-      // Copy msg.data. We take full control of memory in this inline assembly
-      // block because it will not return to Solidity code. We overwrite the
-      // Solidity scratch pad at memory position 0.
-      calldatacopy(0, 0, calldatasize)
-
-      // Call the implementation.
-      // out and outsize are 0 because we don't know the size yet.
-      let result := delegatecall(gas, implementation, 0, calldatasize, 0, 0)
-
-      // Copy the returned data.
-      returndatacopy(0, 0, returndatasize)
-
-      switch result
-      // delegatecall returns 0 on error.
-      case 0 { revert(0, returndatasize) }
-      default { return(0, returndatasize) }
+      switch success
+        // delegatecall returns 0 on error.
+        case 0 { revert(add(data, 32), returndatasize) }
+        default { return(add(data, 32), returndatasize) }
```

Comment:
A small optimization, although the compiler may catch this too.

```suggestion
        let data_start := add(data, 32)
        case 0 { revert(data_start, returndatasize) }
        default { return(data_start, returndatasize) }
```

I also considered using `mload(data)` instead of `returndatasize` but it seems to be more expensive.

Reply:
I tried `mload(data)`, but tests failed

Reply:
That's interesting. Did you understand why?

Reply:
@frangio not yet, will check